src/bin/basic_subclass.rs

```diff
@@ -0,0 +1,87 @@
+//! # Basic Subclass example
+//!
+//! This sample demonstrates how to create a toplevel `window`, set its title, size and
+//! position, how to add a `button` to this `window` and how to connect signals with
+//! actions.
+
+#[macro_use]
+extern crate glib;
+extern crate gio;
+extern crate gtk;
+
+use gio::prelude::*;
+use gtk::prelude::*;
+
+use glib::subclass;
+use glib::subclass::prelude::*;
+use glib::translate::*;
+
+use gio::ApplicationFlags;
+
+#[derive(Debug)]
+struct State {
+    some_window: gtk::Window,
+    some_label: gtk::Label,
+}
+
+pub struct SimpleApplicationPrivate {
+    state: RefCell<Option<State>>,
+}
+
+impl ObjectSubclass for SimpleApplicationPrivate {
+    const NAME: &'static str = "SimpleApplicationPrivate";
+    type ParentType = gtk::Application;
+    type Instance = subclass::simple::InstanceStruct<Self>;
+    type Class = subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn new() -> Self {
+        Self { state: RefCell::new(None), }
+    }
+}
+
+impl ObjectImpl for SimpleApplicationPrivate {
+    glib_object_impl!();
+}
+
+impl gio::subclass::prelude::ApplicationImpl for SimpleApplicationPrivate {
+    fn activate(&self, app: &gio::Application) {
+        let app = app.downcast_ref::<gtk::Application>().unwrap();
+        let window = gtk::ApplicationWindow::new(app);
+        window.present();
+        window.show_all();
+        println!("activated!");
+    }
+}
+
+impl gtk::subclass::application::ApplicationImpl for SimpleApplicationPrivate {}
+
+glib_wrapper! {
+    pub struct SimpleApplication(Object<subclass::simple::InstanceStruct<SimpleApplicationPrivate>, subclass::simple::ClassStruct<SimpleApplicationPrivate>, SimpleApplicationClass>) @extends gio::Application, gtk::Application;
+
+    match fn {
+        get_type => || SimpleApplicationPrivate::get_type().to_glib(),
+    }
+}
+
+fn main() {
+    let app = glib::Object::new(
+        SimpleApplication::static_type(),
+        &[
+            ("application-id", &"org.gtk-rs.SimpleApplication"),
+            ("flags", &ApplicationFlags::empty()),
+        ],
+    )
+    .unwrap()
+    .downcast::<SimpleApplication>()
+    .unwrap();
```

Comment:
IMHO this better be function `new(application_id: &str)` in SimpleApplication

Reply:
Also why not use `.and_then` instead first `unwrap` and `expect` instead second?

Reply:
`new(application_id: &str)`

It doesn't make sense that you would use more than 1 app-id or that you would instantiate more than 1 instance of the application. This never happens and I am not even sure if its possible so I don't see any compelling reason to abstract this.

> Also why not use .and_then instead first unwrap and expect instead second?

Should probably be using `.expect` instead, but no point in using and_then. These kind of errors are fatal and should never fail in the first place. They are like asserts really.

Reply:
For cleanness, moving it into a `new()` function would make sense I guess, but not really important. I agree about the assertions :)